### ISSUE TITLE
Flag to indicate that an external transaction wraps each .sql file

### ISSUE THREAD

**Issue by andrewsmith** [sbdchd/squawk#260]
Hello again! I was working through how to use Squawk to lint migrations at my company. We write SQL files but use a migration tool to apply them to the database. This tool wraps each SQL file in a transaction unless we tell it not to. I think this is somewhat common in SQL migration tools from my experience and research. Because of that, the SQL files don't have BEGIN / COMMIT statements in them. Squawk doesn't have a way of knowing that currently. Rules like `prefer-robust-stmts` warn about problems that aren't really problems due to the external transaction. Others are affected too like `require-concurrent-index-creation`, where the index and the table are created in the same file, but neither is in a transaction according to Squawk.

Initially, I was thinking about writing a script that wrapped each SQL file with BEGIN and COMMIT statements before passing them to Squawk. Squawk would recognize that these were run in a transaction and warnings would be more accurate. The downside would be that line numbers would no longer be accurate in error messages, since this would prepend a line for the BEGIN statement. I could see that being confusing for our developers.

So instead, I was thinking that perhaps Squawk could have an additional flag like `--with-external-transaction` that indicates that a migration SQL file will be wrapped in a transaction when executed. This would then be passed around to various rules and used when determining whether a rule was violated.

I'm happy to create a PR for this if it seems desirable. At a first glance, it is probably a fairly extensive change, but the project is relatively small too.

**Comment by sbdchd** [comment]
I think there was a similar request a while ago and I think it’s best to manually add the begin commit statements to the sql before sending it off to squawk

https://github.com/sbdchd/squawk/issues/106

**Comment by andrewsmith** [comment]
Steve, thanks for your response! I apologize for not noticing that: I looked through open issues and scanned the closed ones, but missed it.

In our case, we have ~3000 migrations: it's a lot to change and for no benefit since our migration tool already wraps those in a transaction. I don't see us making permanent changes to the migrations, many of which are years old now. That's where I went down the path of temporarily wrapping each file with BEGIN/COMMIT statements before passing it to Squawk. If we do something hacky like not add a newline after the "BEGIN; " statement, the line numbers should be unaffected. The column and what's printed in the error might be awkward if the statement with an error is on the first line of the file. I think that is feasible though and we'll reconsider it too.

However, it seems like this is at least somewhat common in SQL migration tools. Here are some examples that I have found:
* [Flyway](https://documentation.red-gate.com/fd/execute-in-transaction-186679827.html)
* [Yoyo Migrations](https://ollycope.com/software/yoyo/latest/#transactions)
* [dbmate](https://github.com/amacneil/dbmate), includes the down migration in the same SQL file, which is awkward for Squawk
* [sqlmigrate](https://github.com/rubenv/sql-migrate)
* [Rails' Migrations](https://edgeguides.rubyonrails.org/active_record_migrations.html), while not SQL-only out of the box, it lets you execute arbitrary SQL which can be read from files

These tools appear to expect explicit transactions in SQL files:
* [Sqitch](https://sqitch.org/)
* Feeding migration files straight into psql, e.g. homegrown tools

A bit of a mix, I guess.

I see the point about making rules more complex. Some of that can be mitigated by grouping concepts into a context struct that gets passed to each rule function.

I'm midway through working on this and happy to share this if there is any chance it would be accepted.

**Comment by sbdchd** [comment]
> The column and what's printed in the error might be awkward if the statement with an error is on the first line of the file.

yeah that error reporting lining up with the code is what I worry about, but I'm open to it if we can get it working without the rules being effected



also re, running migrations in transactions, I'm curious how those tools differentiate between their default (wrapping in a transaction) and migrations where they can't wrap in a transaction?


Like presumably you have some migrations that can't be run in a transaction because they use `concurrently` and related functionality

**Comment by andrewsmith** [comment]
I appreciate your time in responding to my thoughts!

> also re, running migrations in transactions, I'm curious how those tools differentiate between their default (wrapping in a transaction) and migrations where they can't wrap in a transaction?

I know for tools like Yoyo Migrations, there's a special comment you put in your SQL file to indicate that it should be non-transactional. For example:
```sql
-- transactional: false
```

Very similar for dbmate and sqlmigrate.

It looks like Flyway does some [~~dark magic~~ parsing and using Database-specific knowledge](https://documentation.red-gate.com/fd/migrations-184127470.html#Migrations-Transactions) to determine whether a migration may be run in a transaction.

In our case, we're using a variation of Rails migrations to load SQL files and the corresponding *.rb has `disable_ddl_transaction!` in it.

Trying to understand each tool's way of flagging whether a migration should be run in a transaction or not seems unreasonable for Squawk to do. But allowing users to wrap Squawk with a script that partitions SQL files into transactional and non-transactional and calls Squawk with the appropriate flag seems plausible. At least to me 😄 

> edit: further context, the way I've typically used squawk is with Django migrations which has a sqlmigrate option to dump the generated SQL, then I have a script that runs this over all the migrations in a PR and squawk looks at the generated files

Ah, that makes sense, thanks for explaining.

**Comment by sbdchd** [comment]
yeah it seems common enough for it to makes sense to include the `wrap in transaction` or whatever we want to call it

**Comment by aldenquimby** [comment]
Confirming this is now fixed via #264 ?

**Comment by chdsbd** [comment]
@aldenquimby Yes, it seems fixed in squawk-cli@0.19.0
